src/javascript/dimensions-metrics-explorer/components/columnGroupList.tsx

```diff
@@ -0,0 +1,346 @@
+  cubes: CubesByColumn;
+  columns: Column[];
+}> = ({ allowDeprecated, searchTerms, onlySegments, columns, cubes }) => {
+  // Get the grouped columns
+  const groupedColumns = React.useMemo(() => {
+    let filteredColumns: Column[] = columns;
+
+    if (!allowDeprecated) {
+      filteredColumns = filteredColumns.filter(
+        column => column.attributes.status != "DEPRECATED"
+      );
+    }
+
+    if (onlySegments) {
+      filteredColumns = filteredColumns.filter(
+        column => column.attributes.allowedInSegments === "true"
+      );
+    }
+
+    filteredColumns = filteredColumns.filter(column =>
+      searchTerms.every(
+        term =>
+          column.id.toLowerCase().indexOf(term) != -1 ||
+          column.attributes.uiName.toLowerCase().indexOf(term) != -1
+      )
+    );
+
+    // JS Sets guarantee insertion order is preserved, which is important
+    // because the key order in this groupBy determines the order that
+    // they appear in the UI.
+    return groupBy(filteredColumns, column => column.attributes.group);
+  }, [columns, searchTerms, allowDeprecated, onlySegments]);
+
+  // Set of column groups that are currently expanded
+  const [open, setOpen] = React.useState<Set<string>>(() => Set());
+
+  // Expand/Collapse callbacks
+  const toggleGroupOpen = React.useCallback(
+    (group: string) =>
+      setOpen(oldOpen =>
+        oldOpen.contains(group) ? oldOpen.remove(group) : oldOpen.add(group)
+      ),
+    [setOpen]
+  );
+
+  const collapseAll = React.useCallback(() => setOpen(Set()), [setOpen]);
+
+  const expandAll = React.useCallback(
+    () => setOpen(Set.fromKeys(groupedColumns)),
+    [setOpen, groupedColumns]
+  );
+
+  // When a search term is entered, auto-expand all groups. When the search
+  // term is empty, auto-collapse all groups.
+  // KNOWN BUG: this effect should only run when searchTerms changes,
+  // but currently it also runs when expandAll changes, which happens
+  // when a checkbox is checked. Not sure how to fix this. Maybe a ref?
+  /*
+  React.useEffect(() => {
+    if (searchTerms.length === 0) {
+      collapseAll();
+    } else {
+      expandAll();
+    }
+  }, [searchTerms.length === 0, collapseAll, expandAll]);
+  */
+
+  // selectedColumns is a dictionary mapping each column to its selected
+  // state. Each column is associated with one or more "cubes", and a
+  // only columns that share cubes may be mutually selected.
+  const [selectedColumns, setSelectedColumns] = React.useState<Set<string>>(
+    Set
+  );
+
+  const selectColumn = React.useCallback(
+    (column: string, selected: boolean) =>
+      setSelectedColumns(oldSelected =>
+        selected ? oldSelected.add(column) : oldSelected.remove(column)
+      ),
+    [setSelectedColumns]
+  );
+
+  // The set of allowable cubes. When any columns are selected, the set of
+  // allowable cubes is the intersection of the cubes for those columns
+  const allowableCubes = React.useMemo<Set<string> | null>(
```

Comment:
instead of | null, couldn't the code just take empty set to mean there are no allowable Cubes?

Reply:
No, because it breaks intersections. If no columns are selected, then *all* cubes are allowable

Reply:
Although I realized now I can just use an `allCubes` set